### ISSUE TITLE
UTF-8 issues

### ISSUE THREAD

**Issue by ludwich** [pytube/pytube#11]
Im having some problem with your code using letters like å ä ö in it. just try to download a movie named "hypnositören".

**Comment by nficano** [comment]
Can you please provide me a link to the video so I can ensure that I'm accurately reproducing it?

**Comment by ludwich** [comment]
http://www.youtube.com/watch?v=4vGgAk2NDO8

**Comment by rbrcsk** [comment]
[Here's one](https://www.youtube.com/watch?v=kZPaDp3Id8U). It can be only reproduced on 2.X.  The Title of the video is not decoded correctly. I broke it with [this](https://github.com/NFicano/python-youtube-download/commit/99721455bfe5689d76cbcefa8be1a4d62fd974bd) commit.

Removing `.decode()` from the end of [this](https://github.com/NFicano/python-youtube-download/blob/master/pytube/api.py#L225) line and the `__future__` imports would fix it, but break python 3 compatibility.
I suppose we could check if the python version is 3.x and only then decode it. That's the only thing I can think of as I don't have much experience with unicode stuff.

**Comment by ludwich** [comment]
Thats true. Just tried it on python 3 and that worked but my server must have 2.X =( Iw would be sweet it this could be fixed with the check ver =).

**Comment by nanonyme** [comment]
The decode looks wrong anyway though. The decode most likely should actually be using .decode(content_type) where content_type is extracted from headers instead of .decode(). The current code breaks if local encoding isn't the same as Youtube's encoding.

**Comment by rbrcsk** [comment]
@nanonyme Thanks, I forgot about this. The headers don't contain information about the encoding though. These are the headers:

``` python
{'cache-control': 'no-cache',
 'connection': 'close',
 'content-type': 'application/x-www-form-urlencoded',
 'date': 'Mon, 29 Jul 2013 18:33:04 GMT',
 'expires': 'Tue, 27 Apr 1971 19:44:06 EST',
 'p3p': 'CP="This is not a P3P policy! See http://support.google.com/accounts/bin/answer.py?answer=151657&hl=en-US for more info."',
 'server': 'gwiseguy/2.0',
 'set-cookie': 'VISITOR_INFO1_LIVE=u-aXg1wRWDc; path=/; domain=.youtube.com; expires=Sun, 30-Mar-2014 06:26:04 GMT, YSC=rLzKXn2xGbw; path=/; domain=.youtube.com; httponly',
 'transfer-encoding': 'chunked',
 'x-content-type-options': 'nosniff',
 'x-frame-options': 'SAMEORIGIN',
 'x-xss-protection': '1; mode=block',
 'x-youtube-other-cookies': 'VISITOR_INFO1_LIVE=u-aXg1wRWDc;PREF=f1=50000000'}
```

**Comment by gotme** [comment]
In this example, the name of the video uses the symbol for registered trademark.  This funny character, ®, causes a problem.  Other examples are the ellipsis, hyphen, dash and the left/right quotation marks.
https://www.youtube.com/watch?v=Yyc07tnnpQk

My fix was to identify all those funny characters and convert them, as follows.  Hopefully there's a more generic way of doing this.
  filename = yt.filename
  filename = filename.replace('\u2019', '\'')
  filename = filename.replace('\u2018', '\'')
  filename = filename.replace('\u2026', '...')
  filename = filename.replace('\xb2', '')
  filename = filename.replace('\u2014', '-')
  filename = filename.replace('\xa3', '#')
  filename = filename.replace('\xae', '(R)')
  yt.filename = filename

**Comment by gotme** [comment]
Here's a real simple HACK to convert all those high ASCII characters to low ASCII characters, plus any additional multi-byte Unicode converts that you require.

replacements=[
['\u2019','\''],['\u2018','\''],['\u2026','...'],['\u2014','-'],
['\x80','{euro}'],['\x81',''],['\x82','\''],['\x83','f'],['\x84','"'],['\x85','...'],['\x86','+'],['\x87','++'],['\x88','^'],['\x89','{mille}'],
['\x8a','S'],['\x8b','['],['\x8c','OE'],['\x8d',''],['\x8e','Z'],['\x8f',''],['\x90',''],['\x91','\''],['\x92','\''],['\x93','"'],
['\x94','"'],['\x95','(.)'],['\x96','-'],['\x97','--'],['\x98','~'],['\x99','(tm)'],['\x9a','S'],['\x9b',']'],['\x9c','oe'],['\x9d',''],
['\x9e','z'],['\x9f','Y'],['\xa0',''],['\xa1','!'],['\xa2','{cent}'],['\xa3','{pound}'],['\xa4','{currency}'],['\xa5','{yen}'],['\xa6','!'],['\xa7','{section}'],
['\xa8','-'],['\xa9','(c)'],['\xaa','a'],['\xab','[['],['\xac','{not}'],['\xad',''],['\xae','(R)'],['\xaf','_'],['\xb0','{degrees}'],['\xb1','{+-}'],
['\xb2','^2'],['\xb3','^3'],['\xb4','^'],['\xb5','u'],['\xb6','{p}'],['\xb7','.'],['\xb8','_'],['\xb9','^1'],['\xba','O'],['\xbb',']]'],
['\xbc','1/4'],['\xbd','1/2'],['\xbe','3/4'],['\xbf','?'],['\xc0','A'],['\xc1','A'],['\xc2','A'],['\xc3','A'],['\xc4','A'],['\xc5','A'],
['\xc6','AE'],['\xc7','C'],['\xc8','E'],['\xc9','E'],['\xca','E'],['\xcb','E'],['\xcc','I'],['\xcd','I'],['\xce','I'],['\xcf','I'],
['\xd0','D'],['\xd1','N'],['\xd2','O'],['\xd3','O'],['\xd4','O'],['\xd5','O'],['\xd6','O'],['\xd7','x'],['\xd8','O'],['\xd9','U'],
['\xda','U'],['\xdb','U'],['\xdc','U'],['\xdd','Y'],['\xde','b'],['\xdf','b'],['\xe0','a'],['\xe1','a'],['\xe2','a'],['\xe3','a'],
['\xe4','a'],['\xe5','a'],['\xe6','ae'],['\xe7','c'],['\xe8','e'],['\xe9','e'],['\xea','e'],['\xeb','e'],['\xec','i'],['\xed','i'],
['\xee','i'],['\xef','i'],['\xf0','6'],['\xf1','n'],['\xf2','o'],['\xf3','o'],['\xf4','o'],['\xf5','o'],['\xf6','o'],['\xf7','{divided-by}'],
['\xf8','o'],['\xf9','u'],['\xfa','u'],['\xfb','u'],['\xfc','u'],['\xfd','y'],['\xfe','b'],['\xff','y'] ]

filename = yt.filename
for charFrom, charTo in replacements:
  filename = filename.replace(charFrom, charTo)
yt.filename = filename

**Comment by nanonyme** [comment]
@randomchars 
try:
    foo = foo.decode(encoding)
except AttributeError:
    pass # Python3

# curl to a video:

HTTP/1.1 200 OK
Date: Wed, 25 Mar 2015 15:29:26 GMT
Pragma: no-cache
Expires: Fri, 01 Jan 1990 00:00:00 GMT
Cache-Control: no-cache, must-revalidate
Content-Type: text/html; charset=ISO-8859-1
Server: gws
X-XSS-Protection: 1; mode=block
Alternate-Protocol: 443:quic,p=0.5
Transfer-Encoding: chunked
Accept-Ranges: none

# Vary: Accept-Encoding

# curl to main page

HTTP/1.1 200 OK
Date: Wed, 25 Mar 2015 15:30:46 GMT
Server: gwiseguy/2.0
Cache-Control: no-cache
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block; report=https://www.google.com/appserve/security
-bugs/log/youtube
P3P: CP="This is not a P3P policy! See http://support.google.com/accounts/answer
/151657?hl=fi for more info."
Content-Type: text/html; charset=utf-8
X-Frame-Options: SAMEORIGIN
Expires: Tue, 27 Apr 1971 19:44:06 EST
Set-Cookie: YSC=7pbMC3QcqDU; path=/; domain=.youtube.com; HttpOnly
Set-Cookie: VISITOR_INFO1_LIVE=ZMDosiVffmo; expires=Tue, 24-Nov-2015 03:23:46 GM
T; path=/; domain=.youtube.com; HttpOnly
Set-Cookie: PREF=f1=50000000; expires=Tue, 24-Nov-2015 03:23:46 GMT; path=/; dom
ain=.youtube.com
Alternate-Protocol: 443:quic,p=0.5
Transfer-Encoding: chunked
Accept-Ranges: none

# Vary: Accept-Encoding

So Youtube definitely sends charset in Content-Type. Not sure why you're not seeing it. It also varies from page to page so you actually need to extract it. Alternatively just set requests as a dependency. You can get guaranteed unicode objects from its responses that are decoded properly. You also get a HTTP library that works with both Python2 and Python3 without hacking required so it's pretty much win-win.

**Comment by jbwincek** [comment]
Pull request #62 to fix issue #61 also fixed this issue right? So it can now be closed.

**Comment by joaoricardo000** [comment]
Just tried here with version 0.3.0 and had no problems. I this it's closed.
